packages/app-form-builder/src/hooks/useFormDragAndDrop.ts

```diff
@@ -0,0 +1,135 @@
+import { DropTarget, DropSource, DropDestination, FbFormModelField, FbFormStep } from "~/types";
+
+interface UseFormDragParams {
+    editField: (field: FbFormModelField | null) => void;
+    setDropDestination: (desitnation: DropDestination | null) => void;
+}
+
+interface HadleDropParams {
+    target: DropTarget;
+    source: DropSource;
+    destination: DropDestination;
+}
+
+interface CreateCustomFieldParams {
+    data: FbFormModelField;
+    dropDestination: DropDestination;
+}
+
+interface ComposeHadleDropParams {
+    item: DragObjectWithFieldInfo;
+    destinationPosition: DropDestination["position"];
+    formStep: FbFormStep;
+}
+
+export const useFormDragAndDrop = (params: UseFormDragParams) => {
+    const { data, moveRow, moveField, getFieldPlugin, insertField } = useFormEditor();
+
+    const { editField, setDropDestination } = params;
+
+    const handleDrop = useCallback(
+        (params: HadleDropParams) => {
+            const { target, source, destination } = params;
+
+            if (target.name === "custom") {
+                // We can cast because field is empty in the start.
+                editField({} as FbFormModelField);
+                setDropDestination(destination);
+                return;
+            }
+
+            if (target.type === "row") {
+                // Reorder rows.
+                // Reorder logic is different depending on the source and destination position.
+                // "source" is a container from which we move row.
+                // "destination" is a container in which we move row.
+                moveRow(source.position.row, destination.position.row, source, destination);
+                return;
+            }
+
+            if (source.position) {
+                if (source.position.index === null) {
+                    console.log("Tried to move Form Field but its position index is null.");
+                    console.log(source);
+                    return;
+                }
+                const sourceContainer = data.steps.find(step => step.id === source.containerId);
+                const fieldId = sourceContainer?.layout[source.position.row][source.position.index];
+                if (!fieldId) {
+                    console.log("Missing data when moving field.");
+                    return;
+                }
+                moveField({ field: fieldId, target, source, destination });
+                return;
+            }
+
+            // Find field plugin which handles the dropped field type "name".
+            const plugin = getFieldPlugin({ name: target.name });
+            if (!plugin) {
+                return;
+            }
+            insertField({
+                data: plugin.field.createField(),
+                target,
+                destination
+            });
+        },
+        [data]
+    );
+
+    const composeHandleDropParams = (params: ComposeHadleDropParams) => {
```

Comment:
Can we think of a better naming here?


I mean, this is still a drop handler, so the name should reflect that.

Maybe we can just name it `handleDrop` if the original `onDrop` is not being used anywhere.

Reply:
Ignore this comment.